### PR TITLE
Bert

### PR DESCRIPTION
**Pull request by Zenglinxiao**
Add BERT into OpenNMT-py.

### CODE REVIEW

File: docs/source/FAQ.md [hunk 1]
```diff
@@ -136,3 +136,146 @@ E.g.
 will mean that we'll look for `my_data.train_A.*.pt` and `my_data.train_B.*.pt`, and that when building batches, we'll take 1 example from corpus A, then 7 examples from corpus B, and so on.
 
 **Warning**: This means that we'll load as many shards as we have `-data_ids`, in order to produce batches containing data from every corpus. It may be a good idea to reduce the `-shard_size` at preprocessing.
+
+## How do I use BERT?
+BERT is a general-purpose "language understanding" model introduced by Google, it can be used for various downstream NLP tasks and easily adapted into a new task using transfer learning. Using BERT has two stages: Pre-training and fine-tuning. But as the Pre-training is super expensive, we do not recommande you to pre-train a BERT from scratch. Instead loading weights from a existing pretrained model and fine-tuning it is suggested. Currently we support sentence(-pair) classification and token tagging downstream task.
```
Comment (francoishernandez) [inline docs/source/FAQ.md]: "recommande" --> "recommand"

File: docs/source/FAQ.md [hunk 2]
```diff
@@ -136,3 +136,146 @@ E.g.
...
+
+### Use pretrained BERT weights
+To use weights from a existing huggingface's pretrained model, we provide you a script to convert huggingface's BERT model weights into ours.
+
+Usage:
+```bash
+bert_ckp_convert.py  --layers NUMBER_LAYER
```
Comment (francoishernandez) [inline docs/source/FAQ.md]: `python bert_ckp_convert.py ...`

File: docs/source/FAQ.md [hunk 3]
```diff
@@ -136,3 +136,146 @@ E.g.
...
+                     --bert_model_weights_file HUGGINGFACE_BERT_WEIGHTS
+                     --output_name OUTPUT_FILE
+```
+* Go to modeling_bert.py in https://github.com/huggingface/pytorch-transformers/ to check all available pretrained model.
```
Comment (francoishernandez) [inline docs/source/FAQ.md]: `[modeling_bert.py](https://github.com/huggingface/pytorch-transformers/)`

File: docs/source/FAQ.md [hunk 4]
```diff
@@ -136,3 +136,146 @@ E.g.
...
+
+### Preprocess train/dev dataset
+To genenrate train/dev data for BERT, you can use preprocess_bert.py by providing raw data in certain format and choose a BERT Tokenizer model `-vm` coherent with pretrained model.
```
Comment (francoishernandez) [inline docs/source/FAQ.md]: "genenrate" --> "generate"

File: bert_ckp_convert.py [hunk 5]
```diff
@@ -0,0 +1,146 @@
...
+                                   % hugface_key)
+            if 'generator' not in key:
+                onmt_key = re.sub(r'bert\.(.*)', r'\1', key)
+                model_weights['bert'][onmt_key] = weights[hugface_key]
+            else:
+                onmt_key = re.sub(r'generator\.(.*)', r'\1', key)
+                model_weights['generator'][onmt_key] = weights[hugface_key]
+    except ValueError:
+        print("Unsuccessful convert!")
+        exit()
```
Comment (pltrdy) [inline bert_ckp_convert.py]: It's not really explicit + `exit()` quit without trace which isn't helping

File: docs/source/FAQ.md [hunk 6]
```diff
@@ -136,3 +136,146 @@ E.g.
...
+
+* For csv file, each line should contain a instance with one or two sentence column and one column for label as in GLUE dataset, other csv format dataset should be compatible. A typical csv file should be like:
+
+  | ID | SENTENCE_A               | SENTENCE_B(Optional)      | LABEL   |
+  | -- | ------------------------ | ------------------------  | ------- |
+  | 0  | sentence a of instance 0 | sentence b of instance 0  | class 2 |
+  | 1  | sentence a of instance 0 | sentence b of instance 1  | class 1 |
+  | ...| ... | ...  | ... |
+
+  Then calling preprocess_bert.py and providing input sentence columns and label column:
```
Comment (francoishernandez) [inline docs/source/FAQ.md]: `preprocess_bert.py` (backtick around script name)

File: docs/source/FAQ.md [hunk 7]
```diff
@@ -136,3 +136,146 @@ E.g.
...
+                            -vm bert-base-cased --max_seq_len 256 [--do_lower_case]
+                            [--sort_label_vocab] [--do_shuffle]
+  ```
+* For plain text format, we accept multiply files as input, each file contains instances for one specific class. Each line of the file contain one instance which could be composed by one sentence or two separated by ` ||| `. All input file should be arranged in following way:
+  ```
+  .../LABEL_1/filename
+  .../LABEL_2/filename
+  .../LABEL_3/filename
+  ```
+  Then call preprocess_bert.py as following to generate training data:
```
Comment (francoishernandez) [inline docs/source/FAQ.md]: backticks

File: docs/source/FAQ.md [hunk 8]
```diff
@@ -136,3 +136,146 @@ E.g.
...
+  ```bash
+  python preprocess_bert.py --task classification --corpus_type {train, valid}
+                            --file_type csv [--delimiter '\t'] [--skip_head]
+                            --input_columns 1 2 --label_column 3
+                            --data DATA_DIR/FILENAME.tsv
+                            --save_data dataset
+                            -vm bert-base-cased --max_seq_len 256 [--do_lower_case]
+                            [--sort_label_vocab] [--do_shuffle]
+  ```
+* For plain text format, we accept multiply files as input, each file contains instances for one specific class. Each line of the file contain one instance which could be composed by one sentence or two separated by ` ||| `. All input file should be arranged in following way:
```
Comment (francoishernandez) [inline docs/source/FAQ.md]: "contain" --> "contains"
Also, not sure to understand the way "All input file should be arranged".

File: docs/source/FAQ.md [hunk 9]
```diff
@@ -136,3 +136,146 @@ E.g.
...
+                            --data DIR_BASE/LABEL_1/FILENAME1 ... DIR_BASE/LABEL_N/FILENAME2
+                            --save_data dataset
+                            --vocab_model {bert-base-uncased,...}
+                            --max_seq_len 256 [--do_lower_case]
+                            [--sort_label_vocab] [--do_shuffle]
+  ```
+#### Tagging
+For tagging dataset, we support input file in plain text file format.
+
+Each line of the input file should contain token and its tagging, different fields should be separated by a delimiter(default space) while sentences are separated by a blank line.
```
Comment (francoishernandez) [inline docs/source/FAQ.md]: "token" --> "one token" ?

File: docs/source/FAQ.md [hunk 10]
```diff
@@ -136,3 +136,146 @@ E.g.
...
+                            --data DATA_DIR/FILENAME
+                            --save_data dataset
+                            --vocab_model {bert-base-uncased,...}
+                            --max_seq_len 256 [--do_lower_case]
+                            [--sort_label_vocab] [--do_shuffle]
+  ```
+#### Pretraining objective
+Even if it's not recommended, we also provide you a script to generate pretraining dataset as you may want to finetuning a existing pretrained model on masked language modeling and next sentence prediction.
+
+The script expect a single file as input, consisting of untokenized text, with one sentence per line, and one blank line between documents.
```
Comment (francoishernandez) [inline docs/source/FAQ.md]: expects

File: docs/source/FAQ.md [hunk 11]
```diff
@@ -136,3 +136,146 @@ E.g.
...
+                                          [--do_lower_case] [--do_whole_word_mask] [--reduce_memory]
+                                          --epochs_to_generate 2
+                                          --max_seq_len 128
+                                          --short_seq_prob 0.1 --masked_lm_prob 0.15
+                                          --max_predictions_per_seq 20
+                                          [--save_json]
+```
+
+### Training
+After preprocessed data have been generated, you can load weights from a pretrained BERT and transfer it to downstream task with a task specific output head. This task specific head will be initialized by a method you choose if there is no such architecture in weights file specified by `--train_from`. Among all available optimizer, you are suggest to use `--optim bertadam` as it is the method used to train BERT. `warmup_steps` could be set as 1% of `train_steps` as in original paper if use linear decay method.
```
Comment (francoishernandez) [inline docs/source/FAQ.md]: optimizers
suggested

File: docs/source/FAQ.md [hunk 12]
```diff
@@ -136,3 +136,146 @@ E.g.
...
+
+A usage example is given below:
+```bash
+python3 train.py --is_bert --task_type {pretraining, classification, tagging}
```
Comment (francoishernandez) [inline docs/source/FAQ.md]: python - if you want to be consistent with the rest

File: docs/source/FAQ.md [hunk 13]
```diff
@@ -136,3 +136,146 @@ E.g.
...
+                 --learning_rate 2e-5 --learning_rate_decay 0.99 --decay_method linear
+                 --train_steps 4000 --valid_steps 200 --warmup_steps 40
+                 [--report_every 10] [--seed 3435]
+                 [--tensorboard] [--tensorboard_log_dir LOGDIR]
+```
+
+### Predicting
+After training, you can use `predict.py` to generate predicting for raw file. Make sure to use the same BERT Tokenizer model `--vocab_model` as in training data.
+
+For classification task, file to be predicting should be one sentence(-pair) a line with ` ||| ` separating sentence.
```
Comment (francoishernandez) [inline docs/source/FAQ.md]: predicting --> predicted

File: onmt/train_single.py [hunk 14]
```diff
@@ -51,41 +51,64 @@ def main(opt, device_id, batch_queue=None, semaphore=None):
...
-            f_iter = iter(f)
-        except TypeError:
-            f_iter = [(side, f)]
-        for sn, sf in f_iter:
-            if sf.use_vocab:
-                logger.info(' * %s vocab size = %d' % (sn, len(sf.vocab)))
+    if opt.is_bert:
+        # Report bert tokens vocab sizes, including for features
+        f = fields['tokens']
+        if f.use_vocab:  # NOTE: useless!
```
Comment (pltrdy) [inline onmt/train_single.py]: should it be removed then?

File: onmt/train_single.py [hunk 15]
```diff
@@ -51,41 +51,64 @@ def main(opt, device_id, batch_queue=None, semaphore=None):
...
+                    logger.info(' * %s vocab size = %d' % (sn, len(sf.vocab)))
 
     # Build model.
-    model = build_model(model_opt, opt, fields, checkpoint)
-    n_params, enc, dec = _tally_parameters(model)
-    logger.info('encoder: %d' % enc)
-    logger.info('decoder: %d' % dec)
-    logger.info('* number of parameters: %d' % n_params)
+    if opt.is_bert:
+        model = build_bert_model(model_opt, opt, fields, checkpoint)  # V2
```
Comment (pltrdy) [inline onmt/train_single.py]: `# V2`?

File: onmt/trainer.py [hunk 16]
```diff
@@ -148,7 +159,7 @@ def _accum_count(self, step):
                 _accum = self.accum_count_l[i]
         return _accum
 
-    def _maybe_update_dropout(self, step):
+    def _maybe_update_dropout(self, step):  # TODO: to be test with Bert
```
Comment (pltrdy) [inline onmt/trainer.py]: should it (`# TODO`) be done before merging?

File: onmt/train_single.py [hunk 17]
```diff
@@ -51,41 +51,64 @@ def main(opt, device_id, batch_queue=None, semaphore=None):
...
+            vocab = torch.load(opt.data + '.vocab.pt')
     else:
         checkpoint = None
         model_opt = opt
         vocab = torch.load(opt.data + '.vocab.pt')
 
     # check for code where vocab is saved instead of fields
     # (in the future this will be done in a smarter way)
-    if old_style_vocab(vocab):
+    if opt.is_bert:  # TODO: test amelioration
```
Comment (pltrdy) [inline onmt/train_single.py]: should it (`TODO`) be done before merging?

File: onmt/translate/predictor.py [hunk 18]
```diff
@@ -0,0 +1,337 @@
...
+import onmt.model_builder
+import onmt.inputters as inputters
+from onmt.utils.misc import set_random_seed
+
+
+def build_classifier(opt, logger=None, out_file=None):
+    """Return a classifier with result redirect to `out_file`."""
+
+    if out_file is None:
+        out_file = codecs.open(opt.output, 'w+', 'utf-8')
```
Comment (pltrdy) [inline onmt/translate/predictor.py]: (a detail, but wasn't `w` enough here?)

File: onmt/translate/predictor.py [hunk 19]
```diff
@@ -0,0 +1,337 @@
...
+        logger=logger
+    )
+    return classifier
+
+
+def build_tagger(opt, logger=None, out_file=None):
+    """Return a tagger with result redirect to `out_file`."""
+
+    if out_file is None:
+        out_file = codecs.open(opt.output, 'w+', 'utf-8')
```
Comment (pltrdy) [inline onmt/translate/predictor.py]: (a detail, but wasn't w enough here?)

File: onmt/translate/predictor.py [hunk 20]
```diff
@@ -0,0 +1,337 @@
...
+            model,
+            fields,
+            gpu=opt.gpu,
+            verbose=opt.verbose,
+            out_file=out_file,
+            logger=logger,
+            seed=opt.seed)
+
+    def _log(self, msg):
+        if self.logger:
```
Comment (pltrdy) [inline onmt/translate/predictor.py]: You are checking whether logger is `None` or not right? Let's do it explicitly `if self.logger is None:`

File: onmt/translate/predictor.py (hunk 18 above)
Comment (Zenglinxiao) [reply onmt/translate/predictor.py]: Ah, yes you're right. I just copied the same block in translator.py.

**Comment by vince62s** [comment]
@Zenglinxiao don't you think it could be possible to embed "bert_build_model" within "build_model" ?

**Comment by rajarsheem** [comment]
Do you have some comparison between BERT and w/o BERT on NMT tasks?

**Comment by francoishernandez** [comment]
@rajarsheem this is not the aim of BERT or this PR. You have some examples of what it's for in the `FAQ.md`.

File: onmt/encoders/bert.py [hunk 21]
```diff
@@ -0,0 +1,144 @@
...
+        """
+
+        first_token_tensor = hidden_states[:, 0, :]  # [batch, d_model]
+        pooled_output = self.activation_fn(self.dense(first_token_tensor))
+        return pooled_output
+
+
+class BertLayerNorm(nn.Module):
+    def __init__(self, hidden_size, eps=1e-12):
+        """Layernorm module in the TF style(epsilon inside the square root).
```
Comment (guillaumekln) [inline onmt/encoders/bert.py]: Does PyTorch implement it differently?

Comment (Zenglinxiao) [reply onmt/encoders/bert.py]: I adapt the BertLayerNorm from huggingface/pytorch-transformers as it indicate "epsilon inside the square root", and i checked Pytorch doc, it use eps=1e-5, so i kept that. Thanks to your remind,  I find it's the same as i go deep into pytorch cpp code. The origin implementation is due to the typo of Pytorch doc which is fix already (https://github.com/huggingface/pytorch-transformers/pull/1089).
I'll switch that!

Comment (Zenglinxiao) [reply onmt/encoders/bert.py]: BertLayerNorm switched to pytorch original LayerNorm.

File: onmt/model_builder.py [hunk 22]
```diff
@@ -191,40 +213,62 @@ def build_base_model(model_opt, fields, gpu, checkpoint=None, gpu_id=None):
         pad_idx = tgt_base_field.vocab.stoi[tgt_base_field.pad_token]
         generator = CopyGenerator(model_opt.dec_rnn_size, vocab_size, pad_idx)
 
+    if model_opt.is_bert:
+        model = encoder
```
Comment (kagrze) [inline onmt/model_builder.py]: Why the encoder becomes the whole model? If there is no decoder, then how can I use the model for machine translation?

Comment (francoishernandez) [reply onmt/model_builder.py]: BERT is Bidirectional Encoder Representations from Transformers. It is not a machine translation model.
https://arxiv.org/abs/1810.04805

Comment (kagrze) [reply onmt/model_builder.py]: I know this :) But this is a PR to a machine translation system. If the machine translation cannot benefit from this PR then why bother? If you want use BERT for basic downstream tasks (e.g. classification), then you can use Hugging Face's Transformers.

Comment (francoishernandez) [reply onmt/model_builder.py]: OpenNMT-py is not only machine translation, there are other tasks such as Speech2Text, Image2Text, Vid2Text, etc.
This PR is meant to add BERT task(s) and components to the framework to allow users to experiment. Nothing stops you to try and build a custom model using BERT pre-trained encoder layers and standard Transformer decoder layers for instance, but I think it has been tried without great benefits.

Comment (kagrze) [reply onmt/model_builder.py]: Hmm, but AFAIK, all the tasks you mentioned (i.e. Speech2Text, Image2Text, Video2Text) are formalised as transduction problems and, therefore, require some sort of a decoder.

Comment (francoishernandez) [reply onmt/model_builder.py]: Sure. All those things are not mutually exclusive.
Feel free to share any research on [the forum](http://forum.opennmt.net/) and eventually PR if you have anything interesting regarding using BERT in a seq2seq task.

**Comment by Simons2017** [comment]
I would like to using BERT in a seq2seq task, but I'm confused about how to preprocess data. Would you like to give me some advice?
